opensearch/src/main/java/org/opensearch/sql/opensearch/executor/OpenSearchQueryManager.java

```diff
@@ -7,33 +7,79 @@
 
 import java.util.Map;
 import lombok.RequiredArgsConstructor;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import org.apache.logging.log4j.ThreadContext;
+import org.opensearch.OpenSearchTimeoutException;
 import org.opensearch.common.unit.TimeValue;
+import org.opensearch.sql.common.setting.Settings;
 import org.opensearch.sql.executor.QueryId;
 import org.opensearch.sql.executor.QueryManager;
 import org.opensearch.sql.executor.execution.AbstractPlan;
+import org.opensearch.threadpool.Scheduler;
 import org.opensearch.threadpool.ThreadPool;
 import org.opensearch.transport.client.node.NodeClient;
 
 /** QueryManager implemented in OpenSearch cluster. */
 @RequiredArgsConstructor
 public class OpenSearchQueryManager implements QueryManager {
 
+  private static final Logger LOG = LogManager.getLogger(OpenSearchQueryManager.class);
+
   private final NodeClient nodeClient;
 
+  private final Settings settings;
+
   public static final String SQL_WORKER_THREAD_POOL_NAME = "sql-worker";
   public static final String SQL_BACKGROUND_THREAD_POOL_NAME = "sql_background_io";
 
   @Override
   public QueryId submit(AbstractPlan queryPlan) {
-    schedule(nodeClient, () -> queryPlan.execute());
+    TimeValue timeout = settings.getSettingValue(Settings.Key.PPL_QUERY_TIMEOUT);
+    schedule(nodeClient, queryPlan::execute, timeout);
 
     return queryPlan.getQueryId();
   }
 
-  private void schedule(NodeClient client, Runnable task) {
+  private void schedule(NodeClient client, Runnable task, TimeValue timeout) {
     ThreadPool threadPool = client.threadPool();
-    threadPool.schedule(withCurrentContext(task), new TimeValue(0), SQL_WORKER_THREAD_POOL_NAME);
+
+    Runnable wrappedTask =
+        withCurrentContext(
+            () -> {
+              final Thread executionThread = Thread.currentThread();
+
+              Scheduler.ScheduledCancellable timeoutTask =
+                  threadPool.schedule(
+                      () -> {
+                        LOG.warn(
+                            "Query execution timed out after {}. Interrupting execution thread.",
+                            timeout);
+                        executionThread.interrupt();
+                      },
+                      timeout,
+                      ThreadPool.Names.GENERIC);
```

Comment:
We can check later if `ScheduledExecutorService` inside `ThreadPool` already provides scheduling task with timeout.